### PR TITLE
Linux support

### LINKED ISSUE #15: Linux Support
It would be nice if Mint would run on Linux too.

(If this is already the case, it would be nice if the read‐me states as much and **it would be wise to have CI run its tests on Linux.**)

### PR DESCRIPTION
**Pull request by yonaskolb**
Resolves #15

### CODE REVIEW

**Comment by yonaskolb** [comment]
Some tests to do with `SwiftCLI.capture` had to be disabled as they would either error or hang on Linux.
Any ideas @jakeheis?

For example this errors:
```swift
let output = try SwiftCLI.capture(packagePath.commandPath.string, "--version")
```

and this just hangs:
```swift
 let tagOutput = try SwiftCLI.capture(bash: "git ls-remote --tags --refs \(packagePath.gitPath)")
```

**Comment by yonaskolb** [comment]
I've re-enabled the failing linux tests (not the hanging ones though) so the failures can be seen in [CircleCI](https://circleci.com/gh/yonaskolb/Mint/78)

File: Sources/MintKit/Mint.swift [hunk 1]
```diff
@@ -279,9 +281,14 @@ public class Mint {
 
     private func buildPackage(name: String, path: Path) throws {
 
-        let osVersion = ProcessInfo.processInfo.operatingSystemVersion
-        let target = "x86_64-apple-macosx\(osVersion.majorVersion).\(osVersion.minorVersion)"
-        let command = "swift build -c release -Xswiftc -target -Xswiftc \(target) -Xswiftc -static-stdlib"
+        #if os(macOS)
+            let target: String
```
Comment (LinusU) [inline Sources/MintKit/Mint.swift]: Is there a reason why this isn't just `let target = ` on line 287?

Comment (yonaskolb) [reply Sources/MintKit/Mint.swift]: Not, not really. This is a holdover from me trying to extract some common parts, but ended up just having 2 completely seperate paths. Good catch

**Comment by jakeheis** [comment]
It looks like `let output = try SwiftCLI.capture(packagePath.commandPath.string, "--version")` is failing because the executable file at `commandPath` is not being marked as executable. When I try to manually execute that command, I get the error:

```shell
> /tmp/mint/packages/github.com_yonaskolb_SimplePackage/build/2.0.0/SimplePackage --version
bash: /tmp/mint/packages/github.com_yonaskolb_SimplePackage/build/2.0.0/SimplePackage: Permission denied
```

Is the `SimplePackage` executable being directly generated by `swift`, or does Mint do some sort of copying/linking of the executable that could potentially lose the executable flag on Linux?

I'll look into the hanging command soon -- not sure why that would be happening

**Comment by yonaskolb** [comment]
Interesting. The executable is generated by `swift build` and then copied over to the mint directory. Maybe there are some security requirements on Linux I'm not aware of

**Comment by jakeheis** [comment]
After investigating the stalling command more, I've found that it's simply because that command generates a ridiculous amount of output, fills up the pipe buffer, then hangs on a `write` call until some of the data is read. Simply replacing

```swift
let taskOut = verbose ? standardOut : PipeStream()
```

with

```swift
let taskOut = verbose ? standardOut : LineStream {_ in}
```
fixes the stall, as the data is constantly being read and thus the pipe never gets full.

**Comment by Dschee** [comment]
Hello guys, it would be really amazing to see Linux support getting finalized as this would make installation of SPM-based Command-Line-Tools much easier to install on Linux.

**Comment by yonaskolb** [comment]
Ok, I've fixed the hangs by using `LineStream {_ in}`. Using `WriteStream.null` actually lead to the following error:
```
Fatal error: POSIX command failed with error: 9 -- EBADF: file Foundation/Process.swift, line 514
```

The capture is still failing though. Can someone test this branch on Linux to see if installs are actually executable?

**Comment by jakeheis** [comment]
For some reason, `NSFileManager.copyItem` seems to remove the executable flag on Linux. Replacing `try toolFile.copy(packagePath.commandPath)` with `try SwiftCLI.run("cp", toolFile.string, packagePath.commandPath.string)` in `Mint.install` fixes some of the errors.

The remaining errors are due to the build of `SimplePackage` actually failing. After cloning the SimplePackage repo, I successfully ran `swift build` but `swift build -c release -Xswiftc -static-stdlib` (what Mint runs) fails

**Comment by yonaskolb** [comment]
Well we're a bit closer now.
3 remaining issues are:
- linux doesn't support building with `-Xswiftc -static-stdlib`. This means when a new version of swift comes out the old installs won't work on linux. A stable ABI in Swift 5 will remove the need for this as well
- Linux doesn't seem to like executing the symlink we use for global installs
- Due to the casing issues this should really be added first https://github.com/yonaskolb/Mint/issues/17 At the moment Mintfile won't work without it as you can't specify the command name override in a Mintfile

**Comment by Dschee** [comment]
The first of those three issues shouldn't be a problem for using Mint in Continuous Integration services as those will usually install the required dependencies freshly every time a build is run. The second sounds to me like I would need to specify the entire path to the executable, which isn't beautiful but could still get CIs working. I don't understand how the third issues might interfer with using Mint in a CI, maybe it's not a problem if no Mintfile is used?

Would it make sense to merge the work done until now and release a version which might already work in CIs and fix the other stuff in subsequent PRs?

**Comment by yonaskolb** [comment]
Good news is that I got the linked installs working.
I'll merge this now, but may wait till 3 gets resolved to release

**Pull request merged**